core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -474,19 +481,44 @@ private boolean isLocalOnlyLoader(CacheLoader loader) {
 
    @Override
    public void writeToAllStores(MarshalledEntry marshalledEntry, AccessMode mode) {
+      writeToAllNonTxStores(marshalledEntry, mode);
+   }
+
+   @Override
+   public void writeToAllNonTxStores(MarshalledEntry marshalledEntry, AccessMode accessMode) {
       storesMutex.readLock().lock();
       try {
-         for (CacheWriter w : writers) {
-            if (mode.canPerform(configMap.get(w))) {
-               w.write(marshalledEntry);
-            }
+         writers.stream().filter(canPerform(accessMode)).forEach(writer -> writer.write(marshalledEntry));
+      } finally {
+         storesMutex.readLock().unlock();
+      }
+   }
+
+   @Override
+   public void prepareAllTxStores(Transaction transaction, List<MarshalledEntry> modifiedEntries, List<Object> keysToRemove,
+                                  AccessMode accessMode) throws Throwable {
+      storesMutex.readLock().lock();
+      try {
+         for (TransactionalCacheWriter writer : txWriters) {
```

Comment:
Unfortunately we can't re-use ```performOnAllTxStores``` here, because ```prepareWithModifications``` throws Throwable which isn't allowed within a stream.